File: 1642.cpp
#include <queue>

struct Solution {
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        long long required = 0;
        priority_queue<int, vector<int>, std::greater<int>> h;
        for (int i = 0; i < heights.size() - 1; ++i){
            int j = heights[i + 1] - heights[i];
            if (j <= 0){
                continue;
            }
            if (h.size() < ladders){
                h.push(j);
            } else if (ladders == 0 || h.top() >= j){
                required += j;
            } else {
                required += h.top();
                h.pop();
                h.push(j);
            }
            if (required > bricks){
                return i;
            }
        }
        return heights.size() - 1;
    }
};

File: 1642.py
import heapq as hq

class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        h = []
        requiredBricks = 0
        for i in range(len(heights) - 1):
            jump = heights[i + 1] - heights[i]
            if jump <= 0:
                continue
            if len(h) < ladders:
                hq.heappush(h, jump)
            elif ladders == 0 or h[0] >= jump:
                requiredBricks += jump
            else:
                requiredBricks += hq.heapreplace(h, jump)
            if requiredBricks > bricks:
                return i
        return len(heights) - 1

